### DISCUSSION TITLE
Is it possible to take an image as input, and split it out into images of just its colors?

### DISCUSSION THREAD

**Discussion by QuestionableWizard** [ImageMagick/ImageMagick#6638] [Help]
### ImageMagick version

7.1.1-15

### Operating system, version and so on

Mac Monterey

### Description

I included an example image of a smile.
![smile-export](https://github.com/ImageMagick/ImageMagick/assets/144855531/ca1e6007-1bac-4374-b72a-fbac17971f6f)

I want to take that smile as input, and output three PNGs, where each is a separate color in the image. So for that smile it would have a yellow PNG of just the face's background, a blue PNG of the smile, and a black PNG for eyes. The image should stay the same resolution 16x16. If the image had 10 colors, there would be 10 PNGs.

Is this possible with ImageMagick?

**Comment by fmw42** [comment]
Sure.  Use -fuzz XX% +transparent color to make all other colors transparent.  Convert the input to in-memory MPR: format to save from reading multiple times from disk.  Use parenthesis processing to keep each step separate.  Adjust the XX% as needed for your image

```
magick smile.png -write mpr:img +delete \
\( mpr:img -fuzz 25% +transparent blue +write blue.png \) \
\( mpr:img -fuzz 25% +transparent yellow +write yellow.png \) \
\( mpr:img -fuzz 25% +transparent black +write black.png \) \
null:
```

![blue](https://github.com/ImageMagick/ImageMagick/assets/13458218/fe9599c6-c74c-4e96-adbb-027e0ea76561)

![yellow](https://github.com/ImageMagick/ImageMagick/assets/13458218/f5f9f0ad-4943-4726-a096-24c2d25ad34d)

![black](https://github.com/ImageMagick/ImageMagick/assets/13458218/767606ab-4d20-4a80-b8e0-52bf57411147)


You can loop over the histogram to get the list of colors in the image, if there is no anti-aliasing, so that there is a limited number of colors. If you use the histogram colors, then the -fuzz XX% could be as low as 0%. A bash script could parse the information in the histogram to extract just the colors and process the output for each color one-at-a-time, rather than as I did above in one step.

**Comment by QuestionableWizard** [comment]
@fmw42 Thank you for your quick answer. As someone new to ImageMagick, if I try to use portions of that on the command line, for instance the first command, I get:

> magick: missing output filename `+delete' at CLI arg 4 @ error/magick-cli.c/ProcessCommandOptions/753.

There's definitely no anti-aliasing going on in this image, just "pure" colors.

**Reply by fmw42** [reply]
Sorry, My code got munged and lost the backslashes, as I forgot to format it. It is fixed now.  Try again.

**Reply by QuestionableWizard** [reply]
I pasted the following code into the command line. Sorry if I'm being stupid.

```
magick smile.png -write mpr:img +delete 
( mpr:img -fuzz 25% +transparent blue +write blue.png ) 
( mpr:img -fuzz 25% +transparent yellow +write yellow.png ) 
( mpr:img -fuzz 25% +transparent black +write black.png ) 
```

**Comment by GeeMack** [comment]
This command will read an input image, determine how many colors it has, create a duplicate of the image for each unique color, and output an image for each color with everything transparent except that color.

    magick input.png -background none \( +clone -unique-colors \) -duplicate "%[fx:v.w-1],0" \
       -reverse +swap +transparent "%[pixel:u[-1].p{t,0}]" +delete result-%02d.png

If the input image has transparent pixels the output will include an image for that, also.

The output images are named with 2 digits with leading zeros... "result-NN.png".

NOTE: This command could easily create an enormous number of output images depending on the input. To make sure the output is limited to a certain number or fewer, add `-colors N` right after reading in the input image.

I tested the command in Windows and converted it to a *nix command. It should work as written, but I haven't had a chance to try it, so I may have overlooked escaping some special characters.

**Reply by fmw42** [reply]
GeeMack,  works fine on my Mac.  So syntax is fine.  Very nice.
